Add Mohammadi2023Turkiye ML-based GSIM

amirxdbx opened the pull request:
This PR adds a new machine-learning-based Ground Motion Model for Turkey
(Mohammadi et al., 2023) implemented as a GSIM in the OpenQuake hazard library.

Key features:
• Implements the vectorized compute() API (no legacy methods).
• Wraps the official XGBoost models exported to ONNX runtime.
• Includes standard deviations (sigma, tau, phi) from stds.csv.
• Includes verification tables and regression tests.
• Documentation updated in hazardlib.gsim.rst.
• Changelog updated.

The GSIM-specific tests pass successfully.

CB-quakemodel commented:
Dear Amir, 

Many thanks for providing an implementation of your machine-learning based GSIM. I will do my best to review this model by the end of the month. 

One thing I notice immediately, is that there are far too many `.onnx` files added in the current approach. You need to adjust the implementation to store the cached ONNX sessions for each intensity measure within a single file (1 per period is simply too many). Making this change would be much appreciated on our end.

Perhaps @micheles can give some input on this too.

Thanks,

Christopher

CB-quakemodel commented:
> Dear Christopher (@CB-quakemodel),
> 
> Happy New Year, and I apologise for the raised issue. I have bundled all the ONNX files into one and modified the code accordingly. Please check if any other steps are required from my end.
> 
> Bests Amir

Dear @amirxdbx, 

Thanks for making the changes, and a happy new year to yourself also. 

I see many ONNX files are now included in a different location and were not included in the initial commits you made earlier today, so perhaps these are unintentionally remaining in your PR? Also, the workflow_failure.yaml has been modified in your PR, so you should revert this file given it is likely unintentionally modified here.

Thanks,

Christopher

amirxdbx commented:
Dear Christopher, 
Thanks for pointing this out. I removed the extra files and reverted the changes to the yaml file. 
Bests 
Amir

CB-quakemodel commented:
> Dear Christopher, Thanks for pointing this out. I removed the extra files and reverted the changes to the yaml file. Bests Amir

Many thanks! I will complete my review early this week.

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,474 @@
+    if not os.path.exists(_STDS_FILE):
+        raise IOError(f"Cannot find stds.csv at {_STDS_FILE}")
+
+    with open(_STDS_FILE, newline="") as f:
```
For rigour, any external files used within an OQ GSIM must be opened within the GSIM's `__init__` method whenever possible. I see this GSIM currently lacks an `__init__` - this must be added to ensure these CSVs are opened and subsequently managed safely.

Please see https://github.com/gem/oq-engine/blob/master/openquake/hazardlib/gsim/zhao_2016.py#L596 for an example of this.

It is probably not too tricky to move the work done within `_load_stddev_tables` into the required `__init__` method for the GSIM class.

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,474 @@
+    # Run inference only for the requested output
+    out = sess.run([target_output], {input_name: X_scaled})[0]
+    return np.asarray(out, dtype=float).reshape(-1)
+
```
Between functions outside of the class, there should consistently be 2 lines between each function (several other instances of this throughout this file).

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,474 @@
+
+
+# ---------------------------------------------------------------------
+# 4. ONNX model handling (Single bundled model)
```
Each of these numbered points can be moved into the docstring of the functions they correspond to. This will help keep things cleaner and adhere to the 2 lines between each function not within the GSIM class itself.

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,474 @@
+        return "ln(PGV)"
+    if kind == "SA":
+        return f"ln(PSA={period})"
+    raise ValueError(f"Unknown IM kind: {kind}")
```
There are two similar instances of checking if the specified IMT is supported by the GMM (one here, and one in `_mean_and_std_from_ctx`. You probably need just one of these checks given it seems to be  (but please correct me if I am wrong!).

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,474 @@
+    REQUIRES_SITES_PARAMETERS = {"vs30"}
+
+    # ------------------------------------------------------------------
+    # NEW OQ API: vectorized compute()
```
Like my comments above about the numbered points further up in the file (which are really useful for a user such as myself), I think you can move this info into the compute docstring, or just remove it given the docstring already provides similar info).

CB-quakemodel reviewed (changes requested):
Thanks again @amirxdbx, I have made an initial review with only a handful of comments/requested changes (nearly all of them are just to make things cleaner/easier to digest as an external user reviewing the GSIM module). The only moderately big change is to ensure safe handling of any external files used in the GSIM (please see my comments within the review).

amirxdbx commented:
Thanks for the review, Christopher! I’ve updated the code to address your comments.

I moved the documentation into the class strings and added the _init_  method to handle the CSV loading safely as you suggested. I also cleaned up the global variables, fixed the spacing, and consolidated those checks.

Let me know if everything looks good now!

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,397 @@
+    return X * _SCALE + _MIN
+
+
+# ---------------------------------------------------------------------
```
You need to remove lines 99 to 101 to ensure the docs tests pass (more comments like this below)

CB-quakemodel commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,288 @@
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = {PGA, PGV, SA}
+
+    # Component: geometric mean or closest available
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = _IMC_GMEAN
```
This should be fixed to the type (of those permitted in OQ) to whichever is appropriate. Given the _IMC_GMEAN is not specified currently within the init, I guess the fallback of `_IMC_GMEAN = list(const.IMC)[0]` is always being used currently, which selects `GMEAN`.

Therefore, I would remove lines 46 to 61 and specify `DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.GEOMETRIC_MEAN` instead

CB-quakemodel replied:
@amirxdbx  The "docs" CI tests now pass. I have made one final set of comments for my review (just some very small things I did not notice before). Once these final changes are made and the tests all go green (which they should now!), I can approve the PR for merging.

Many thanks for your work on providing this GMM in an OQ implementation!

Christopher

micheles commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,268 @@
+        _SESS_CACHE = ort.InferenceSession(full_path, sess_options=opt)
+
+    return _SESS_CACHE
+
```
We have an unified way of reading .onnx files and it is via the `PicklableInferenceSession`  class, please follow the approach used in `openquake/hazardlib/gsim/taherian_2024_inland.py`, otherwise you are just duplicating logic which is already there.

amirxdbx commented on openquake/hazardlib/gsim/mohammadi_turkiye_2023.py:
```diff
@@ -0,0 +1,268 @@
+
+# ONNX model directory
+_ONNX_DIR = os.path.join(_DATA_DIR, "onnx_models")
+# Standard deviations file
```
I compressed the files into a 9 MB zip file. Unfortunately, reducing the size more than this is not possible at the moment.
